In [1]:
import pandas as pd
import numpy as np

In [2]:
wvs_w6 = pd.read_csv('../data/Evaluating_Happiness/wvs_w6.csv', low_memory=False)
wvs_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6  ...  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  ...   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  ...   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  ...   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  ...   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  ...   

   I_HOMOLIB  I_ABORTLIB  I_DIVORLIB    choice  WEIGHT3B  I_VOICE1  I_VOICE2  \
0        0.0         0.0    0.222222  0.074074       1.0      0.00       0.5   
1        0.0         0.0    0.000000  0.000000       1.0      0.66       0.0   
2        0.0         0.0    0.333333  0.111111       1.0      0.33       0.0   
3        0.0         0.0    0.000000  0.000000       1.0      0.00       0.0   
4        0.0         0.0    0.222222  0.074074       1.0      0.66       0.0   

   I_VOI2_00  voice  WEIGHT4B  
0      0.250  0.250       1.0  
1      0.330  0.330       1.0  
2      0.165  0.165       1.0  
3      0.000  0.000       1.0  
4      0.330  0.330       1.0  

[5 rows x 442 columns]

In [3]:
wvs_w6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89565 entries, 0 to 89564
Columns: 442 entries, V1 to WEIGHT4B
dtypes: float64(428), int64(12), object(2)
memory usage: 302.0+ MB


In [4]:
wvs_w6.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89565 entries, 0 to 89564
Data columns (total 442 columns):
 #   Column           Dtype  
---  ------           -----  
 0   V1               int64  
 1   V2               int64  
 2   V2A              int64  
 3   cow              int64  
 4   C_COW_ALPHA      object 
 5   B_COUNTRY_ALPHA  object 
 6   V3               int64  
 7   V4               float64
 8   V5               float64
 9   V6               float64
 10  V7               float64
 11  V8               float64
 12  V9               float64
 13  V10              float64
 14  V11              float64
 15  V12              float64
 16  V13              float64
 17  V14              float64
 18  V15              float64
 19  V16              float64
 20  V17              float64
 21  V18              float64
 22  V19              float64
 23  V20              float64
 24  V21              float64
 25  V22              float64
 26  V23              float64
 27  V24            

Codebook available here: https://www.worldvaluessurvey.org/WVSDocumentationWV6.jsp (accessed on 5/25/2021)

In [12]:
# how many columns with null values?
wvs_w6.isna().sum().count()

442

In [18]:
# another approach for finding the number of columns with null values
sum(map(any, wvs_w6.isnull()))

442

In [32]:
# how many null values in each column?
pd.set_option('display.max_rows',None)
wvs_w6.isna().sum()

V1                     0
V2                     0
V2A                    0
cow                    0
C_COW_ALPHA            0
B_COUNTRY_ALPHA        0
V3                     0
V4                   341
V5                   545
V6                  1035
V7                  1620
V8                  1519
V9                  1348
V10                  758
V11                  326
V12                   14
V13                    6
V14                    7
V15                   17
V16                    9
V17                   13
V18                   12
V19                   18
V20                   15
V21                   19
V22                    3
V23                  586
V24                 2388
V25                  840
V26                  923
V27                  993
V28                 3211
V29                 3129
V30                 1118
V31                 2223
V32                 1055
V33                 1214
V34                 1197
V35                 9696
V36                 3979


In [16]:
# how many null values in all?
wvs_w6.isna().sum().sum()

10721335

In [27]:
# how many columns contain missing values?
print(np.count_nonzero(wvs_w6.isna().sum() != 0))

# how many columns do not?
print(np.count_nonzero(wvs_w6.isna().sum() == 0))

414
28


#### Observations so far:
This data set contains 28 columns without any missing information, and 414 columns (out of 442) that contain at least 1 missing value.

#### Next steps:
- figure out if these missing values matter for my analyses
- decide what to do about missing values

In [30]:
# list columns containing missing data
wvs_w6.columns[wvs_w6.isna().any()].tolist()

['V4',
 'V5',
 'V6',
 'V7',
 'V8',
 'V9',
 'V10',
 'V11',
 'V12',
 'V13',
 'V14',
 'V15',
 'V16',
 'V17',
 'V18',
 'V19',
 'V20',
 'V21',
 'V22',
 'V23',
 'V24',
 'V25',
 'V26',
 'V27',
 'V28',
 'V29',
 'V30',
 'V31',
 'V32',
 'V33',
 'V34',
 'V35',
 'V36',
 'V37',
 'V38',
 'V39',
 'V40',
 'V41',
 'V42',
 'V43',
 'V44',
 'V44_ES',
 'V45',
 'V46',
 'V47',
 'V48',
 'V49',
 'V50',
 'V51',
 'V52',
 'V53',
 'V54',
 'V55',
 'V56',
 'V56_NZ',
 'V57',
 'V58',
 'V59',
 'V60',
 'V61',
 'V62',
 'V63',
 'V64',
 'V65',
 'V66',
 'V67',
 'V68',
 'V69',
 'V70',
 'V71',
 'V72',
 'V73',
 'V74',
 'V74B',
 'V75',
 'V76',
 'V77',
 'V78',
 'V79',
 'V80',
 'V81',
 'V82',
 'V83',
 'V84',
 'V85',
 'V86',
 'V87',
 'V88',
 'V89',
 'V90',
 'V91',
 'V92',
 'V93',
 'V94',
 'V95',
 'V96',
 'V97',
 'V98',
 'V99',
 'V100',
 'V101',
 'V102',
 'V103',
 'V104',
 'V105',
 'V106',
 'V107',
 'V108',
 'V109',
 'V110',
 'V111',
 'V112',
 'V113',
 'V114',
 'V115',
 'V116',
 'V117',
 'V118',
 'V119',
 'V120',
 'V121',
 'V122',


In [33]:
# Columns to drop
# drop a handful of item scales at the end of the data set since a) these come from items that appear earlier in the survey,
# and b) they don't necessarily reduce the missing values count my much
cleaning_w6 = wvs_w6.drop(['sacsecval',
 'resemaval',
 'I_AUTHORITY',
 'I_NATIONALISM',
 'I_DEVOUT',
 'defiance',
 'I_RELIGIMP',
 'I_RELIGBEL',
 'I_RELIGPRAC',
 'disbelief',
 'I_NORM1',
 'I_NORM2',
 'I_NORM3',
 'relativism',
 'I_TRUSTARMY',
 'I_TRUSTPOLICE',
 'I_TRUSTCOURTS',
 'scepticism',
 'I_INDEP', # autonomy scale items are based on V12, V18, V19, and V21
 'I_IMAGIN', # autonomy scale items are based on V12, V18, V19, and V21
 'I_NONOBED', # autonomy scale items are based on V12, V18, V19, and V21
 'autonomy', # autonomy scale items are based on V12, V18, V19, and V21
 'I_WOMJOB',
 'I_WOMPOL',
 'I_WOMEDU',
 'equality',
 'I_HOMOLIB',
 'I_ABORTLIB',
 'I_DIVORLIB',
 'choice',
 'I_VOICE1',
 'I_VOICE2',
 'I_VOI2_00',
 'voice'], axis=1)

In [ ]:
# 
cleaning_w6 = wvs_w6.drop([ 'MN_35A',
 'MN_163A',
 'MN_163B',
 'MN_163C',
 'MN_228L',
 'MN_228M',
 'MN_228N',
 'MN_228O',
 'MN_228P',
 'MN_228Q',
 'MN_228R',
 'MN_228S1',
 'MN_228S2',
 'MN_228S3',
 'MN_228S4',
 'MN_228S5',
 'MN_228S6',
 'MN_228S7',
 'MN_228S8',
 'MN_229A',
 'MN_229B',
 'MN_230A',
 'MN_233A',
 'MN_233B',
 'MN_234A',
 'MN_237A',
 'MN_237B1',
 'MN_237B2',
 'MN_237B3',
 'MN_237B4',
 'MN_237B5',
 'MN_237B6',
 'MN_237B7',
 'MN_237C1',
 'MN_237C2',
 'MN_237C3',
 'MN_237C4',
 'MN_237C5',
 'MN_237C6',
 'MN_249A1',
 'MN_249A2',
 'MN_249A3']

In [ ]:
pd.DataFrame()

In [31]:
# select columns containing at least one NaN value
#wvs_w6.loc[:, wvs_w6.isna().any()]

V4   V5   V6   V7   V8   V9  V10  V11  V12  V13  ...  I_WOMEDU  \
0      1.0  1.0  1.0  NaN  1.0  1.0  2.0  1.0  1.0  1.0  ...      0.00   
1      1.0  2.0  3.0  4.0  2.0  2.0  2.0  2.0  2.0  1.0  ...      0.33   
2      1.0  3.0  2.0  4.0  2.0  1.0  2.0  2.0  2.0  2.0  ...      0.00   
3      1.0  1.0  3.0  4.0  3.0  1.0  2.0  1.0  2.0  2.0  ...      0.33   
4      1.0  1.0  1.0  2.0  1.0  1.0  1.0  3.0  2.0  1.0  ...      0.66   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...       ...   
89560  1.0  3.0  2.0  4.0  1.0  1.0  3.0  3.0  2.0  1.0  ...      0.66   
89561  1.0  1.0  1.0  3.0  1.0  1.0  3.0  1.0  2.0  1.0  ...      1.00   
89562  1.0  2.0  1.0  3.0  1.0  1.0  3.0  3.0  2.0  1.0  ...      0.66   
89563  1.0  2.0  2.0  3.0  1.0  1.0  2.0  2.0  2.0  1.0  ...      1.00   
89564  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  ...      0.66   

       equality  I_HOMOLIB  I_ABORTLIB  I_DIVORLIB    choice  I_VOICE1  \
0      0.000000        0.0         0.0    0.222222  0.074074      0.00   
1      0.110000        0.0         0.0    0.000000  0.000000      0.66   
2      0.000000        0.0         0.0    0.333333  0.111111      0.33   
3      0.386667        0.0         0.0    0.000000  0.000000      0.00   
4      0.220000        0.0         0.0    0.222222  0.074074      0.66   
...         ...        ...         ...         ...       ...       ...   
89560  0.773333        0.0         0.0    0.000000  0.000000      0.66   
89561  0.553333        0.0         0.0    0.000000  0.000000      0.66   
89562  0.773333        0.0         0.0    0.000000  0.000000      0.33   
89563  1.000000        0.0         0.0    0.444444  0.148148      0.00   
89564  0.606667        0.0         0.0    0.000000  0.000000      0.00   

       I_VOICE2  I_VOI2_00  voice  
0           0.5      0.250  0.250  
1           0.0      0.330  0.330  
2           0.0      0.165  0.165  
3           0.0      0.000  0.000  
4           0.0      0.330  0.330  
...         ...        ...    ...  
89560       1.0      0.830  0.830  
89561       0.5      0.580  0.580  
89562       1.0      0.665  0.665  
89563       1.0      0.500  0.500  
89564       1.0      0.500  0.500  

[89565 rows x 414 columns]

In [14]:
# quick and dirty stats on some columns
wvs_w6.describe()

V1            V2           V2A           cow            V3  \
count  89565.0  89565.000000  89565.000000  89565.000000  8.956500e+04   
mean       6.0    471.966147    486.231943    494.832155  3.994326e+04   
std        0.0    245.114798    251.387109    251.203268  1.918431e+05   
min        6.0     12.000000     12.000000      2.000000  1.000000e+00   
25%        6.0    276.000000    332.000000    349.000000  3.910000e+02   
50%        6.0    434.000000    484.000000    560.000000  8.120000e+02   
75%        6.0    702.000000    710.000000    705.000000  1.356000e+03   
max        6.0    887.000000    901.000000    920.000000  1.394604e+06   

                 V4            V5            V6            V7            V8  \
count  89224.000000  89020.000000  88530.000000  87945.000000  88046.000000   
mean       1.094537      1.673961      1.888693      2.624129      1.516605   
std        0.346876      0.734792      0.837273      0.985910      0.796788   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        1.000000      1.000000      1.000000      2.000000      1.000000   
50%        1.000000      2.000000      2.000000      3.000000      1.000000   
75%        1.000000      2.000000      2.000000      3.000000      2.000000   
max        4.000000      4.000000      4.000000      4.000000      4.000000   

       ...     I_HOMOLIB    I_ABORTLIB    I_DIVORLIB        choice  \
count  ...  82179.000000  84917.000000  86965.000000  85229.000000   
mean   ...      0.252691      0.247608      0.394510      0.297814   
std    ...      0.337607      0.307271      0.343517      0.279625   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.037037   
50%    ...      0.000000      0.111111      0.444444      0.222222   
75%    ...      0.444444      0.444444      0.666667      0.481481   
max    ...      1.000000      1.000000      1.000000      1.000000   

           WEIGHT3B      I_VOICE1      I_VOICE2     I_VOI2_00         voice  \
count  89565.000000  85267.000000  84367.000000  83282.000000  86352.000000   
mean       0.967156      0.333907      0.351476      0.342580      0.343176   
std        0.100441      0.314214      0.376777      0.269469      0.267931   
min        0.660000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000      0.000000      0.000000      0.165000      0.165000   
50%        1.000000      0.330000      0.500000      0.330000      0.330000   
75%        1.000000      0.660000      0.500000      0.580000      0.568960   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

           WEIGHT4B  
count  89565.000000  
mean       0.976149  
std        0.086836  
min        0.660000  
25%        1.000000  
50%        1.000000  
75%        1.000000  
max        1.000000  

[8 rows x 440 columns]